// @gravity-ui/navigation/build/cjs/HotkeysPanel-d971b094.js
'use strict';

var tslib_es6 = require('./tslib.es6-705c6589.js');
var React = require('react');
var uikit = require('@gravity-ui/uikit');
var Drawer = require('./Drawer.js');
var cn = require('./cn-9933321a.js');
var styleInject_es = require('./style-inject.es-dcee06b6.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function filterHotkeys(hotkeys, filter) {
    if (filter) {
        const result = [];
        const normalizedFilter = filter.toLowerCase();
        for (const hotkeysBlock of hotkeys) {
            const filteredItems = hotkeysBlock.items.filter((item) => item.title.toLowerCase().includes(normalizedFilter));
            if (filteredItems.length > 0) {
                result.push(Object.assign(Object.assign({}, hotkeysBlock), { items: filteredItems }));
            }
        }
        return result;
    }
    return hotkeys;
}

function flattenHotkeyGroups(hotkeys) {
    const result = [];
    for (const hotkeysGroup of hotkeys) {
        result.push({
            title: hotkeysGroup.title,
            group: true,
        });
        result.push(...hotkeysGroup.items);
    }
    return result;
}

var css_248z = ".gn-hotkeys-panel{--hotkeys-panel-width:400px;--hotkeys-panel-vertical-padding:18px;--hotkeys-panel-horizontal-padding:24px;inset:0;position:fixed}.gn-hotkeys-panel__drawer-item{box-sizing:border-box;display:flex;flex-direction:column;padding:var(--hotkeys-panel-vertical-padding) 0;width:var(--hotkeys-panel-width)}.gn-hotkeys-panel__title{margin:0 var(--hotkeys-panel-horizontal-padding) 16px var(--hotkeys-panel-horizontal-padding)}.gn-hotkeys-panel__search{box-sizing:border-box;margin-bottom:14px;padding:0 var(--hotkeys-panel-horizontal-padding)}.gn-hotkeys-panel__list{flex:1;overflow-y:auto}.gn-hotkeys-panel__item{height:auto;padding:8px var(--hotkeys-panel-horizontal-padding)}.gn-hotkeys-panel__item.g-list__item_active{background:inherit}.gn-hotkeys-panel__item-content{align-items:baseline;color:var(--g-color-text-primary);display:flex;font-size:var(--g-text-body-1-font-size);justify-content:space-between;line-height:var(--g-text-body-1-line-height);width:100%}.gn-hotkeys-panel__item-content_group{font-size:var(--g-text-body-2-font-size);font-weight:500;line-height:var(--g-text-body-2-line-height)}.gn-hotkeys-panel__hotkey{color:var(--g-color-text-complementary)}";
styleInject_es.styleInject(css_248z);

const b = cn.block('hotkeys-panel');
function HotkeysPanel(_a) {
    var { visible, onClose, leftOffset, topOffset, className, drawerItemClassName, filterClassName, titleClassName, listClassName, itemContentClassName, preventScrollBody, hotkeys, itemClassName, filterable = true, filterPlaceholder, title, emptyState } = _a, listProps = tslib_es6.__rest(_a, ["visible", "onClose", "leftOffset", "topOffset", "className", "drawerItemClassName", "filterClassName", "titleClassName", "listClassName", "itemContentClassName", "preventScrollBody", "hotkeys", "itemClassName", "filterable", "filterPlaceholder", "title", "emptyState"]);
    const [filter, setFilter] = React.useState('');
    const hotkeysList = React.useMemo(() => {
        const filteredHotkeys = filterHotkeys(hotkeys, filter);
        return flattenHotkeyGroups(filteredHotkeys);
    }, [hotkeys, filter]);
    const renderItem = React.useCallback((item) => (React__default["default"].createElement("div", { className: b('item-content', { group: item.group }, itemContentClassName), key: item.title },
        item.title,
        item.value && React__default["default"].createElement(uikit.Hotkey, { className: b('hotkey'), value: item.value }))), []);
    const drawerItemContent = (React__default["default"].createElement(React__default["default"].Fragment, null,
        React__default["default"].createElement("h2", { className: b('title', titleClassName) }, title),
        filterable && (React__default["default"].createElement(uikit.TextInput, { value: filter, onUpdate: setFilter, placeholder: filterPlaceholder, autoFocus: true, className: b('search', filterClassName), hasClear: true })),
        React__default["default"].createElement(uikit.List, Object.assign({ className: b('list', listClassName), virtualized: false, filterable: false, items: hotkeysList, renderItem: renderItem, itemClassName: b('item', itemClassName), emptyPlaceholder: emptyState }, listProps))));
    return (React__default["default"].createElement(Drawer.Drawer, { className: b(null, className), onVeilClick: onClose, onEscape: onClose, preventScrollBody: preventScrollBody, style: {
            left: leftOffset,
            top: topOffset,
        } },
        React__default["default"].createElement(Drawer.DrawerItem, { id: "hotkeys", visible: visible, className: b('drawer-item', drawerItemClassName), content: drawerItemContent })));
}

exports.HotkeysPanel = HotkeysPanel;
