// @gravity-ui/components/build/cjs/components/DefinitionList/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTitle = exports.getAllItemsAsGroups = exports.onlySingleItems = exports.isGroup = exports.isUnbreakableOver = exports.b = void 0;
const cn_1 = require("../utils/cn");
exports.b = (0, cn_1.block)('definition-list');
function isUnbreakableOver(limit) {
    return function (value) {
        const posibleLines = value.split(/\s+/);
        return posibleLines.some((line) => line.length > limit);
    };
}
exports.isUnbreakableOver = isUnbreakableOver;
const isGroup = (item) => 'label' in item && !('name' in item);
exports.isGroup = isGroup;
const onlySingleItems = (items) => !items.some((el) => (0, exports.isGroup)(el));
exports.onlySingleItems = onlySingleItems;
function getAllItemsAsGroups(items) {
    const result = [];
    let temporaryList = [];
    for (const item of items) {
        if ((0, exports.isGroup)(item)) {
            if (temporaryList.length) {
                result.push({ items: temporaryList, label: null });
                temporaryList = [];
            }
            result.push(item);
        }
        else {
            temporaryList.push(item);
        }
    }
    if (temporaryList.length) {
        result.push({ items: temporaryList, label: null });
        temporaryList = [];
    }
    return result;
}
exports.getAllItemsAsGroups = getAllItemsAsGroups;
function getTitle(title, content) {
    if (title) {
        return title;
    }
    if (typeof content === 'string' || typeof content === 'number') {
        return String(content);
    }
    return undefined;
}
exports.getTitle = getTitle;
